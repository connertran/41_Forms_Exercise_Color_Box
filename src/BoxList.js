import { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";
import { v4 as uuid } from "uuid";

function BoxList() {
  const initialState = [];
  const [boxes, setBoxes] = useState(initialState);
  const addBox = (backgroundColor, width, height) => {
    const id = uuid();
    setBoxes((boxes) => [...boxes, { backgroundColor, width, height, id }]);
  };
  const removeBox = (id) => {
    console.log("you clicked");
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };
  return (
    <>
      <h2>Box list</h2>
      <NewBoxForm addBox={addBox} />
      <div>
        {boxes.map(({ id, backgroundColor, width, height }) => (
          <Box
            key={id}
            id={id}
            backgroundColor={backgroundColor}
            width={width}
            height={height}
            removeBox={removeBox}
          />
        ))}
      </div>
    </>
  );
}

export default BoxList;
