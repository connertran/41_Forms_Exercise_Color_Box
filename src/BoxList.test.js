import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
it("renders without crashing", function () {
  render(<BoxList />);
});

// snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add and delete a new item", function () {
  const { getByLabelText, queryByText } = render(<BoxList />);

  // no items yet
  expect(queryByText("X")).not.toBeInTheDocument();

  const backgroundInput = getByLabelText("Background:");
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const addBtn = queryByText("Add box!");

  // fill out the form
  fireEvent.change(backgroundInput, { target: { value: "blue" } });
  fireEvent.change(widthInput, { target: { value: "100px" } });
  fireEvent.change(heightInput, { target: { value: "100px" } });
  fireEvent.click(addBtn);

  // delete button next the box div exists!
  expect(queryByText("X")).toBeInTheDocument();

  //   delete the box
  const deleteBtn = queryByText("X");
  fireEvent.click(deleteBtn);
  //   the box is deleted
  expect(queryByText("X")).not.toBeInTheDocument();
});
