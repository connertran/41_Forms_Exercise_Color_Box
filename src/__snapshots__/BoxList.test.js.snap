// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h2>
    Box list
  </h2>
  <form>
    <label
      for="e64b3cb6-a450-4b28-828e-5de13fb09730"
    >
      Background: 
    </label>
    <input
      id="e64b3cb6-a450-4b28-828e-5de13fb09730"
      name="backgroundColor"
      placeholder=" Enter Background Color"
      type="text"
      value=""
    />
    <label
      for="881175fd-84fa-4b4b-b66e-400300a0211f"
    >
      Width: 
    </label>
    <input
      id="881175fd-84fa-4b4b-b66e-400300a0211f"
      name="width"
      placeholder=" Enter width value, e.g. 140px"
      type="text"
      value=""
    />
    <label
      for="b71a7d32-f098-4c0c-a5ef-a6b305b251f3"
    >
      Height: 
    </label>
    <input
      id="b71a7d32-f098-4c0c-a5ef-a6b305b251f3"
      name="height"
      placeholder=" Enter height value, e.g. 240px"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Add box!
    </button>
  </form>
  <div />
</DocumentFragment>
`;
